---
title: "Module2 Prediction"
author: "Patipark Kueanjinda"
date: "Updated on June 11, 2020"
output: html_notebook
---

```{r}
# call packages
require(tidyverse)
require(tibble)
require(reshape2)
require(randomForest)
require(caret)
```

Download data table.
```{r}
setwd('C:/Users/patipark/Dropbox/CCA manuscript/Final/draft 20/GitData/Patipark/Results/module1_results/Tables')

HT2 <- read.csv('Hallmarks_Matrix_20200611.csv', header = FALSE, stringsAsFactors = FALSE)
colnames(HT2) <- HT2[1,]
colnames(HT2)[c(11,12,13,14)] <- c('leading_edge','','','core_enrichment')
HT2 <- HT2[-(1),] %>% `row.names<-`(., NULL)

# download fold-change expression levels of genes
df.explevel <- read.csv('Expression_level_ALL_20200611.csv', header = FALSE, stringsAsFactors = FALSE)
colnames(df.explevel) <- df.explevel[1,]
df.explevel <- df.explevel[-(1),] %>% `row.names<-`(., NULL)

# download drug response
df.DR <- read.csv('Table_DrugSensitivityClass_20200612.csv', header = FALSE, stringsAsFactors = FALSE)
colnames(df.DR) <- df.DR[1,]
df.DR <- df.DR[-(1),] %>% `row.names<-`(., NULL)
```

```{r}
setwd('C:/Users/patipark/Dropbox/CCA manuscript/Final/draft 20/GitData/Patipark/Data/module1_data')
df.node <- read.csv('SizeofNodeinCCASubtype.csv', header = FALSE, stringsAsFactors = FALSE)
colnames(df.node) <- c('Drug','Subtype1','Subtype2')
df.node <- df.node[-(1),] %>% `row.names<-`(., NULL)
```

Gene selection procedures
1. Retrieve leading-edge genes from significant gene sets
2. Match genes from the gene sets and actual genes in gene expression list

```{r}
# change cell names
# cell lines were arranged in order as shown in Figure 3A
from <- c('KKU.213','KKU.214','KKU.156','HuCCA.1','RBE','KKK.D138','TFK.1','YSCCC','SSP.25','KKU.100','HuCCT.1','KKK.D131','KKU.055','KKK.D068','HuH.28')

to <- c('KKU-213','KKU-214','KKU-156','HuCCA-1','RBE','KKK-D138','TFK-1','YSCCC','SSP-25','KKU-100','HuCCT-1','KKK-D131','KKU-055','KKK-D068','HuH-28')

# build a list of cell lines
map <- setNames(to,from)

# reshape gene expression level table
b.out <- reshape2::melt(df.explevel, id.vars = 'Gene', variable.name = 'Cell', value.name = 'fc.Expr')

```

Subset only CCA cell lines from GSEA report table.
Subset drugs of choice from the list in Figure S4A.
```{r}
# prepare GSEA table
cell <- to
# exclude columns named after CCA cell lines as we are not interested
df.input <- HT2 %>% subset(!HT2$Cell %in% cell)
# get a list of types of comparison (e.g. subgroup1/2, drug sensitive/resistant)
DrugInit <- unique(df.input$Cell) 
# select drugs by node size
# top 10 and bottom 10
# df.node must be arranged according to node size beforehand
total.nodes <- nrow(df.node)
top10 <- c(1:13)
pre10 <- total.nodes-14
bottom10 <- c(pre10:total.nodes)
# sel.drugs <- df.node %>% subset(X %in% c(top10,bottom10))
sel.drugs <- df.node # %>% subset(agent %in% Drug20)
# remove 'meanSTO'
sel.drugs <- sel.drugs[(which(sel.drugs$agent != 'meanSTO')),]
# remove suffix '_S.R' from most drug names
DrugInit <- gsub('_S.R','',DrugInit) 
# create a list of drugs of interest including 'Subtype_1.2' for latter use
A <- DrugInit[which(DrugInit %in% c('Subtype_1.2','OrgSubtype_1.2',sel.drugs$agent))]
DrugInit <- A

```

Now, we wanted to extract gene classifiers.
```{r}
# k = index number of Cell
# DrugInit = list of Cell
# df.input = hallmark gene set table
# df.DR2mod = metadata table
# b.out = gene expression level table

k
DrugInit <- unique(HT2$Cell)
df.input <- HT2
df.DR2mod <- df.DR
df.exp <- b.out %>% subset(Cell %in% to)

# we wrote a function below:
CVV <- function (DrugInit,df.input,df.DR2mod,b.out) {
  k <- 1 # example
  # obtain drug name
  drug.name <- ifelse(DrugInit[k] %in% c('Subtype_1.2','Organ_1.2'), DrugInit[k], DrugInit[k] %>% gsub('_S.R','',.)) %>% trimws(., which = 'both')
  
  # modify drug names in df.input
  df.HT <- df.input %>% subset(Cell %in% drug.name)
  
  # df.input contains 50 gene sets
  # subset data frame by drug (Cell column)
  df.pos <- df.HT %>% subset(Cell %in% dn & NES > 0)
  df.neg <- df.HT %>% subset(Cell %in% dn & NES <= 0)
  
  # From Positive NES, create a table consists of Genes, Hallmarks,NES
  # gene sets are listed by rows
  cp <- lapply(1:nrow(df.pos), function(a) {
    # extract gene in core_enrichment column by row
    gnls <- df.pos$core_enrichment[a]
    # separate gene by string '/'
    n <- strsplit(gnls, '/') # results from strsplit are lists
    dfn <- do.call(cbind,n) # bind lists to a vector
    # create a table of genes extracted from the gene set with annotations
    # :hallmarks, category, cat.color code, NES status (1,2), NES value, comparison type (Cell)
    dfh <- data.frame('Gene' = as.character(dfn),
                      'Hallmark' = as.character(rep(df.pos$ID[a],length(dfn))),
                      'NES' = as.numeric(rep(df.pos$NES[a], length(dfn))),
                      'p.adjust' = as.numeric(rep(df.pos$p.adjust[a], length(dfn))),
                      'qvalues' = as.numeric(rep(df.pos$qvalues[a], length(dfn))),
                      'NES.status' = as.character(rep('positive',length(dfn))),
                      'NES.value' = as.numeric(rep(1,length(dfn))),
                      'Cell' = as.character(rep(dnn,length(dfn))))
  })
  # bind table of all genes with their annotations
  dfp <- do.call(rbind,cp)
  
  # Frome Negative NES, construct a matrix containing Genes, Hallmarks,NES
  cn <- lapply(1:nrow(df.neg), function(b) {
    # extract gene in core_enrichment column by row
    gnls <- df.neg$core_enrichment[b]
    # separate gene by string '/'
    n <- strsplit(gnls, '/') # results from strsplit are lists
    dfn <- do.call(cbind,n) # bind lists to a vector
    # create a table of genes extracted from the gene set with annotations
    # :hallmarks, category, cat.color code, NES status (1,2), NES value, comparison type (Cell)
    dfh <- data.frame('Gene' = as.character(dfn),
                      'Hallmark' = as.character(rep(df.neg$ID[b],length(dfn))),
                      'NES' = as.numeric(rep(df.neg$NES[b], length(dfn))),
                      'p.adjust' = as.numeric(rep(df.neg$p.adjust[b], length(dfn))),
                      'qvalues' = as.numeric(rep(df.neg$qvalues[b], length(dfn))),
                      'NES.status' = as.character(rep('negative',length(dfn))),
                      'NES.value' = as.numeric(rep(2,length(dfn))),
                      'Cell' = as.character(rep(dnn,length(dfn))))
  }) 
  # bind table of all genes with their annotations
  dfn <- do.call(rbind,cn)
  
  # genes selected for further analysis must pass the following criteria:
  # FDR-adjust p-value (p.adjust) < 0.05 to avoid false positive genes
  # Benjamini-Hochberg is more strict to p.adjust
  # moderately strong evidence and null hypothesis does not hold
  dfgene <- rbind(dfp,dfn) %>% subset(p.adjust < 0.05)
  
  # in some cases, no cancer hallmarks was significantly enriched,
  # thus, the p.adjust was set to less stringency
  if(nrow(dfgene) != 0){
    dfgene <- dfgene
  } else {
    # p-value < 0.1 
    # weak evidence and null hypothesis does not hold
    dfgene <- rbind(dfp,dfn) %>% subset(p.adjust < 0.1)
  }
  
  ## create temporary dataframe
  # temp.dfexp <- b.out
  ## extract drug sensitivity
  # sensitivity <- df.DR2mod %>% subset(Drug == dn)
  ## add sensitivity to cell lines
  # temp.dfexp$Class <- sensitivity$Sensitivity[match(temp.dfexp$Cell,sensitivity$Cell)]
  
  ## find which genes are significantly expression in each drug
  # sig.test.gene <- lapply(1:length(unique(b.out$Gene)), function(i) {
  #   gene <- as.character(unique(b.out$Gene))[i]
  #   temp.exp <- b.out %>% subset(Gene == gene)
  #   temp.exp$Cell <- gsub('_S.R','',temp.exp$Cell) %>% trimws(., which = 'both')
  #   temp.exp$Cell <- gsub('\\.','-',temp.exp$Cell)
  #   temp.exp$Cell <- gsub('Subtype_1-2','Subtype_1.2',temp.exp$Cell)
  #   old.name <- c('Caffeic-APE','Mitomycin-C','Mycophenolate-Mofetil','Vinorelbine-tartrate')
  #   new.name <- c('Caffeic APE','Mitomycin C','Mycophenolate Mofetil','Vinorelbine tartrate')
  #   
  #   for(i in 1:length(old.name)){
  #     temp.exp$Cell[which(temp.exp$Cell == old.name[i])] <- new.name[i]
  #   }
  # 
  #   sig.test.drug <- lapply(1:length(unique(df.DR2mod$Drug)), function(j) {
  #     drug <- as.character(unique(df.DR2mod$Drug))[j]
  #     expr <- temp.exp %>% subset(Cell %in% cell)
  #     sensitivity <- df.DR2mod %>% subset(Drug == drug)
  #     A <- merge(expr,sensitivity,by = 'Cell')
  #     A$Sensitivity <- factor(A$Sensitivity, levels = c('Sensitive','Resistant'))
  #     # Ax <- A %>% subset(Sensitivity == 'Sensitive')
  #     # Ay <- A %>% subset(Sensitivity == 'Resistant')
  #     sigvalue <- t.test(exp.fc ~ Sensitivity,
  #                        data = A, 
  #                        var.equal = FALSE)$p.value
  #     output <- data.frame('Drug' = drug,
  #                          'pvalue' = as.numeric(sigvalue))
  #     return(output)
  #   })
  #   sigByDrug <- do.call(rbind,sig.test.drug)
  #   
  # })
  
  # siggenes <- temp.dfexp %>% group_by(Gene) %>% summarise(sigvalue = sig.test(Sensitive,Resistant) )
  
  ## create two sets of cells
  # if it's CCA subtype
  if(drug.name == 'Subtype_1.2') {
    my.Sall <- df.DR2mod %>% group_by(Cell,CCA.Subgroup) %>% summarise()
    colnames(my.Sall) <- c('Cell','Sensitivity')
    my.S1 <- my.Sall %>% subset(Sensitivity == 'Sensitive')
    my.S2 <- my.Sall %>% subset(Sensitivity == 'Resistant')
    # if it's Organ subtype
  } else if (drug.name == 'Organ_1.2') {
    my.Sall <- df.DR2mod %>% group_by(Cell,Organ.Subgroup) %>% summarise()
    colnames(my.Sall) <- c('Cell','Sensitivity')
    my.S1 <- my.Sall %>% subset(Sensitivity == 'Sensitive')
    my.S2 <- my.Sall %>% subset(Sensitivity == 'Resistant')
    # if it's individual drug response subtype
  } else {
    my.S1 <- df.DR2mod %>% subset(Drug == dnn & Class.KDE == 'Sensitive') %>% dplyr::select(Cell,Class.KDE)
    colnames(my.S1) <- c('Cell','Sensitivity')
    my.S2 <- df.DR2mod %>% subset(Drug == dnn & Class.KDE == 'Resistant') %>% dplyr::select(Cell,Class.KDE)
    colnames(my.S2) <- c('Cell','Sensitivity')
    # put them to gether for later use
    my.Sall <- rbind(my.S1,my.S2)
  }
  
  # create an input table for selection of predictive genes using random forest model
  # using fold-change of expression levels of cell lines compared to population's median
  dx <- dcast(df.exp, Cell ~ Gene, value.var = 'fc.Expr')
  
  # matching by Cell, add drug Sensitivity
  dx$Sensitivity <- my.Sall$Sensitivity[match(dx$Cell,my.Sall$Cell)]
  
  ### Random sampling 100-times
  # define number of fold to be performed
  fold <- 100
  # a list of cells in sensitive and resistant group
  S1 <- my.S1$Cell
  S2 <- my.S2$Cell
  
  # check number of cell in each group
  no.S1 <- length(S1)
  no.S2 <- length(S2)
  
  # each group must consist of at least 3 cell lines
  set.seed(100)
  if (no.S2 < 3) {
    # down sampling S1
    rnd.sample.S1 <- lapply(1:fold, function(n) sample(S1, 3, replace = FALSE))
    # keep S2
    rnd.sample.S2 <- lapply(1:fold, function(n) sample(S2, no.S2, replace = FALSE))
  } else if (no.S1 < 3) {
    # down sampling S2
    rnd.sample.S2 <- lapply(1:fold, function(n) sample(S2, 3, replace = FALSE))
    # keep S1
    rnd.sample.S1 <- lapply(1:fold, function(n) sample(S1, no.S1, replace = FALSE))
  } else {
    rnd.sample.S1 <- lapply(1:fold, function(n) sample(S1, 3, replace = FALSE))
    # random sample 77% from S2
    rnd.sample.S2 <- lapply(1:fold, function(n) sample(S2, 3, replace = FALSE))
  }
  # create a list of random sampled Cells
  rnd.sample.list <- lapply(1:fold, function(n) {
    
    c(rnd.sample.S1[[n]],rnd.sample.S2[[n]])
    
  })
  
  # Next, we generated gene expression levels from list of cell lines above.
  
  ### In this step, gene expression level would be used to predict class of cell lines (n = 6)
  # idead taken from https://stackoverflow.com/questions/47960427/how-to-calculate-the-oob-of-random-forest
  # loop would run 100 times
  cvrf.out <- lapply(1:fold, function(f) { #5 corresponds to the number of folds defined earlier
    require(dplyr)
    # f <- 1
    # create x matrix of expression levels from list of cell lines
    dataset <- dx %>% subset(Cell %in% c(rnd.sample.list[[f]])) %>% `rownames<-`(., NULL) %>% 
      tibble::column_to_rownames(., var = 'Cell')
    
    # count number of sensitive and resistant cells
    no.S <- dataset %>% subset(Sensitivity == 'Sensitive') %>% nrow()
    no.R <- dataset %>% subset(Sensitivity == 'Resistant') %>% nrow()
    # if there was 1 cell in the group, we added two repeated of the single row
    if(no.S == 1) {
      df <- dataset %>% subset(Sensitivity == 'Sensitive')
      df <- df[rep(seq_len(nrow(df)), each = 2),]
      dataset <- rbind(dataset, df)
    } else if(no.R == 1) {
      df <- dataset %>% subset(Sensitivity == 'Resistant')
      df <- df[rep(seq_len(nrow(df)), each = 2),]
      dataset <- rbind(dataset, df)
    } else {
      dataset <- dataset
    }
    # add factor levels for random forest function
    dataset$Sensitivity <- factor(dataset$Sensitivity, levels = c('Sensitive','Resistant'))
  
    # build a control for training model
    require(caret)
    set.seed(100)
    control <- caret::trainControl(method = 'LOOCV', 
                                   number = 10,
                                   # repeats = 10,
                                   search = 'grid',
                                   savePredictions = 'final',
                                   classProbs = TRUE, 
                                   summaryFunction = twoClassSummary,
                                   selectionFunction = 'best',
                                   sampling = 'down',
                                   allowParallel = TRUE)

    # build random forest model
    train.rf <- caret::train(Sensitivity ~ .,
                           data = dataset,
                           method = 'rf', 
                           metric = 'ROC',
                           # tuneGrid = tunegrid,
                           trControl = control,
                           na.omit = FALSE,
                           importance = TRUE)
    
    ## build SVM train model
    # train.svmRadial <- caret::train(Sensitivity ~ .,
    #                                 data = dataset,
    #                                 method = 'svmRadial',
    #                                 metric = 'ROC',
    #                                 trControl = control)
    ## build boosted classification tree model
    # train.ada <- caret::train(Sensitivity ~ .,
    #                           data = dataset,
    #                           method = 'ada',
    #                           metric = 'ROC',
    #                           trControl = control)
    ## build generalized linear model 
    # train.glm <- caret::train(Sensitivity ~ .,
    #                                data = dataset,
    #                                method = 'glm',
    #                                metric = 'ROC',
    #                                trControl = control)

    # using sample dataset, determine which algorithm is the best.
    # models_compare <- caret::resamples(list(
    #   RF = train.rf, SVM = train.svmRadial
    # ))
    # the models comparison showed that random forest is better than SVM in term of specificity and sensitivity.
    # therefore, we use random forest as the algorithm tob build classification model.
    
    testset <- dx %>% subset(!Cell %in% c(rnd.sample.list[[f]])) %>% `rownames<-`(., NULL) %>%
      tibble::column_to_rownames(., var = 'Cell')
    
    # check testset data, at least 1 sample should be positive
    idxS <- which(testset$Sensitivity == 'Sensitive')
    idxR <- which(testset$Sensitivity == 'Resistant')
    
    if(is_empty(idxS)) {
      # get 1 row of positive Cell
      addrow <- dx %>% subset(Cell %in% c(rnd.sample.list[[1]])) %>% subset(Sensitivity == 'Sensitive') %>%
        sample_n(., 1) %>% `rownames<-`(., NULL) %>% tibble::column_to_rownames(., var = 'Cell')
      testset <- rbind(testset,addrow)
    } else if (is_empty(idxR)) {
      addrow <- dx %>% subset(Cell %in% c(rnd.sample.list[[1]])) %>% subset(Sensitivity == 'Resistant') %>%
        sample_n(., 1) %>% `rownames<-`(., NULL) %>% tibble::column_to_rownames(., var = 'Cell')
      testset <- rbind(testset,addrow)
    } else {
      testset <- testset
    }

    # add factor level
    testset$Sensitivity <- factor(testset$Sensitivity, levels = c('Sensitive','Resistant'))


    predset <- predict(train.rf,type = 'prob', testset[,-(which(colnames(testset) == 'Sensitivity'))])
    # add observed class
    predset$observe <- testset$Sensitivity[match(rownames(predset),rownames(testset))]
    
    predset$observe <- factor(predset$observe, levels = c('Sensitive','Resistant'))
        
    AUC <- pROC::roc(observe ~ Sensitive, data = predset)$auc
    
    # accuracy <- caret::confusionMatrix(predset, as.factor(testset$Sensitivity))
    # ba <- accuracy$byClass[[11]]

    
    # 
    # outputs <- data.frame('Drug' = ,
    #                       'Model' = j,
    #                       'Round' = as.numeric(ba))
    
    ### Add criteria for stringent selection of genes
    # we use two approaches to obtain useful variables
    # 1. obtain variables from random forest model
    varimp.rf <- caret::varImp(train.rf, scale = TRUE)
    varimp.table <- as.data.frame(varimp.rf$importance)
    
    # filter variable using varimp > 80
    varimp.table <- varimp.table %>% subset(Sensitive > 80)
    
    # 2. calculate area under ROC of each variable, regardless of the random forest model
    roc_imp <- caret::filterVarImp(x = dataset[, -ncol(dataset)],
                            y = dataset$Sensitivity,
                            nonpara = TRUE)
    # filter variable using a cut-off point of AUC = 1 for excellent classifiers.
    varimp.roc <- roc_imp %>% subset(Sensitive == 1)

    # find intersect genes
    sel.genes <- intersect(rownames(varimp.roc), rownames(varimp.table))

    # use to selected variables from ROC to subset variables from random forest model
    varimp.table2 <- varimp.table %>% subset(rownames(.) %in% sel.genes) #%>%
      #subset(Sensitive >= 80)
    varimp.roc <- varimp.roc %>% subset(rownames(.) %in% sel.genes)

    # ### prediction
    # # Cells <- c(S1,S2)
    # # Cells.ex <- !Cells %in% rnd.sample.list[[1]]
    # # Cells.ex <- Cells[Cells.ex]
    # # dataset.ex <- dx %>% subset(Cell %in% Cells.ex) %>% `rownames<-`(., NULL) %>% 
    # #   tibble::column_to_rownames(., var = 'Cell')
    # # dataset.ex$Sensitivity <- factor(dataset.ex$Sensitivity, levels = c('Sensitive','Resistant'))
    # # 
    # # testData <- predict(train.rf, dataset.ex[,-ncol(dataset.ex)], type = 'prob')
    # # pred <- ifelse(testData[, 'Sensitive'] > 0.5, 'Sensitive', 'Resistant')
    # # pred <- factor(pred, levels = c('Sensitive','Resistant'))
    # # # evaluate accuracy using confusionMatrix
    # # my.cf <- confusionMatrix(reference = dataset.ex$Sensitivity,
    # #                 data = pred,
    # #                 mode = 'everything',
    # #                  positive = 'Sensitive')
    # ###
    # 
    outputs <- data.frame('Gene' = rownames(varimp.table2),
                          'Drug' = rep(dn,nrow(varimp.table2)),
                          'test.ROC.AUC' = rep(as.numeric(AUC),nrow(varimp.table2)),
                          # 'train.ROC.AUC' = varimp.roc$Sensitive,
                          'Varimp.Score' = varimp.table2$Sensitive)
    return(outputs)
  })
  
  # Build a list of genes that pass the criteria
  CV_VIP <- do.call(rbind,cvrf.out)
  
  if(is.null(CV_VIP) != TRUE) {
    table.out <- CV_VIP %>% subset(test.ROC.AUC >= 0.8)
    table.out <- table.out[!duplicated(table.out$Gene),] %>% as.data.frame()
    # add drug name
    table.out$Drug <- rep(dn,nrow(table.out))
    
    
  } else {
    
    table.out <- data.frame('Gene' = NA,
                            'Drug' = dn,
                            'test.ROC.AUC' = paste0('< 0.8'),
                            # 'train.ROC.AUC' = paste0('< 0.8'),
                            'Varimp.Score' = NA
                            )
    
  }
  return(table.out)
}

Biomarkers.test.organ <- lapply(2:2,function(x) {CVV(x,DrugInit,df.input,df.DR2mod,b.out)})

# Biomarkers.test.33 <- table.out
Biomarkers.test.list <- rbind(
  Biomarkers.test.1[[1]],
  Biomarkers.test.2[[1]],
  Biomarkers.test.3[[1]],
  Biomarkers.test.4[[1]],
  Biomarkers.test.5[[1]],
  Biomarkers.test.6[[1]],
  Biomarkers.test.7[[1]],
  Biomarkers.test.8[[1]],
  Biomarkers.test.9[[1]],
  Biomarkers.test.10[[1]],
  do.call(rbind,Biomarkers.test.1120),
  do.call(rbind,Biomarkers.test.2130),
  Biomarkers.test.31[[1]],
  Biomarkers.test.32[[1]],
  Biomarkers.test.33,
  Biomarkers.test.34[[1]],
  Biomarkers.test.35[[1]],
  do.call(rbind,Biomarkers.test.3640),
  do.call(rbind,Biomarkers.test.4150),
  do.call(rbind,Biomarkers.test.5160),
  do.call(rbind,Biomarkers.test.6170),
  do.call(rbind,Biomarkers.test.7181)
)

Biomarkers.test.org <- do.call(rbind,Biomarkers.test.organ) 
colnames(Biomarkers.test.org) <- c('Gene','Drug','ROC.Score','Varimp.Score')
Biomarkers.test.org$Drug <- gsub('OrgSubtype_1.2','Organ Subgroup',Biomarkers.test.org$Drug)
Biomarkers.test.list <- rbind(Biomarkers.test.list,Biomarkers.test.org)

# test2 <- CVV(1)
# 
# Biomarkers.test.list <- Biomarkers.test.list %>% subset(!Drug == 'Subtype_1.2')
# 
# Biomarkers.test.list <- rbind(test2, Biomarkers.test.list)

```

Save table for selected genes
```{r}
setwd('C:/Users/patipark/Dropbox/CCA project/CCA RNA-seq/NewGeneSet_Module_InitialAnalysis/prediction_function')
write.csv(Biomarkers.test.list, file = 'PredictedGenes_20191111.csv', quote = FALSE, row.names = FALSE)
# Biomarkers.test.list <- read.csv('PredictedGenes_11072019_v2.csv',header = TRUE,stringsAsFactors = TRUE)
```

Evaluate ROC AUC of the 4 models using CCA cell line dataset
```{r}
require(ggplot2)
require(plotROC)

# create table from train model of each classifier
train.rf$pred$model <- rep('RF',8)
train.svmRadial$pred$model <- rep('SVM Radial',8)
train.glm$pred$model <- rep('GLM',8)
train.ada$pred$model <- rep('ADA',8)

sel.cols <- c('model','pred','obs','Sensitive','Resistant')
my.rf <- train.rf$pred %>% dplyr::select(sel.cols)
my.svm <- train.svmRadial$pred %>% dplyr::select(sel.cols)
my.glm <- train.glm$pred %>% dplyr::select(sel.cols)
my.ada <- train.ada$pred %>% dplyr::select(sel.cols)

train.all <- rbind(my.rf, my.svm, my.glm, my.ada)

train.all$model <- factor(train.all$model, levels = unique(train.all$model))

train.ROC <- ggplot(train.all, 
                    aes(m = Sensitive, 
                        d = factor(obs, levels = c("Sensitive", "Resistant")))) + 
  geom_roc(hjust = -0.4, vjust = 1.5, labelsize = 0,pointsize = 0.1) + 
  geom_abline(intercept = 0, 
              slope = 1, 
              linetype = 'dashed', 
              color = 'gray', 
              size = 0.2) +
  facet_grid(. ~ model) +
  coord_equal() +
  theme_minimal() +
  theme(
    text = element_text(size = 12),
    axis.text = element_text(size = 6),
    axis.title = element_text(size = 8)
  ) +
  labs(x = 'False Positive Rate \n(1-Specificity)',
       y = 'True Positive Rate \n(Sensitivity)')


train.ROC.AUC <- calc_auc(train.ROC)
train.ROC.AUC$model <- c('RF','SVM Radial','GLM','ADA') %>% as.factor()

train.all$AUC <- train.ROC.AUC$AUC[match(train.all$model,train.ROC.AUC$PANEL)]

p <- train.ROC + geom_text(
  data    = train.ROC.AUC,
  mapping = aes(label = paste0('AUC = ',as.character(AUC))),
  size = 2,
  x = 0.4,
  y = 0.2,
  hjust   = 0,
  vjust   = 0,
  inherit.aes = FALSE
) 

```

Save ROC Figure S5A
```{r}
setwd('C:/Users/patipark/Dropbox/CCA project/CCA RNA-seq/NewGeneSet_Module_InitialAnalysis/prediction_function')
# ggsave(p,
#        filename = 'train_models_ROC.pdf',
#        height = 5,
#        width = 8,
#        units = 'cm',
#        device = cairo_pdf
#        )

pdf('train_models_ROC.pdf',
    # paper = 'A4',
    width = 6, #inches
    height = 11, #inches
    useDingbats = FALSE  #avoid letters instead of point objects
)
p

dev.off()
```

End prediction_function script here


Figure 4B
Draw overlapping genes across drugs
```{r}
require(tidyverse)
# extract leading edges from significant gene sets (p.adjust < 0.05)
B <- lapply(1:length(DrugInit), function(j) {
  # j <- 1
  # extract drug name
  dn <- DrugInit[j]
  # subset enrichment table by drug
  df.input$Cell <- gsub('_S.R','',df.input$Cell) %>% trimws(., which = 'both')
  enrich.table <- df.input %>% subset(Cell == DrugInit[j])
  # only gene sets with significant level threshold < 0.05 will be selected
  dfgene <- enrich.table %>% subset(p.adjust < 0.05)
  # if dataframe is NULL, adjust significant level threshold to < 0.1
  if(nrow(dfgene) != 0){
    dfgene <- dfgene
  } else {
    # p-value < 0.1 
    # weak evidence and null hypothesis does not hold
    dfgene <- enrich.table %>% subset(p.adjust < 0.1)
  }
  # extrace genes from significantly enriched hallmarks
  A <- lapply(1:nrow(dfgene), function(i) {
    # get core_enrichment column
    ss <- dfgene$core_enrichment[i]
    # split string list
    ss2 <- strsplit(ss,split = '/')
    ss.df <- data.frame('Gene' = do.call(cbind,ss2))
    Hallmark <- rep(dfgene$ID[i],nrow(ss.df))
    # append Hallmark column
    ss.df <- tibble::add_column(ss.df, Hallmark, .after = 1)
    return(ss.df)
  })
  # build dataframe
  AA <- do.call(rbind,A)
  # append Drug name
  Drug <- rep(DrugInit[j], nrow(AA))
  AA <- tibble::add_column(AA, Drug, .after = 3)
  return(AA)
})

# Table of leading-edge genes from significant cancer hallmarks
BB <- do.call(rbind,B)
```

Extract gene expression level
```{r}
mat.output <- df.explevel
# modify cell names in the columns of the table
from <- c('KKU.213','KKU.214','KKU.156','HuCCA.1','RBE','KKK.D138','TFK.1','YSCCC','SSP.25','KKU.100','HuCCT.1','KKK.D131','KKU.055','KKK.D068','HuH.28')

to <- c('KKU-213','KKU-214','KKU-156','HuCCA-1','RBE','KKK-D138','TFK-1','YSCCC','SSP-25','KKU-100','HuCCT-1','KKK-D131','KKU-055','KKK-D068','HuH-28')

map <- setNames(to,from)

# rename cell names in df.explevel
ln.col <- length(colnames(df.explevel))
b <- lapply(1:ln.col, function(n) {
  idx.name <- which(colnames(df.explevel) %in% names(map))
  if (n %in% idx.name){
    a <- as.data.frame(df.explevel[,c(n)])
    colnames(a)[1] <- map[which(names(map) == colnames(df.explevel)[n])]
  } else {
    a <- as.data.frame(df.explevel[,c(n)])
    colnames(a)[1] <- colnames(df.explevel)[n]
  }
  return(a)
})
# Reshape fold-change mRNA expression level and subset only CCA cells
dat.expr <- do.call(cbind,b) %>% reshape2::melt(., id.vars = 'Gene', 
                                             variable.name = 'Cell', 
                                             value.name = 'expr') %>% 
  subset(Cell %in% c(to,from))

dat.expr <- dat.expr %>% group_by(Gene,Cell) %>% summarise(expr = mean(expr))

# add drug sensitivity defined by 100-drug response and individual drug
S1 <- c('KKK-D138','HuCCA-1','KKU-214','RBE','KKU-156','KKU-213')
S2 <- c('KKK-D131','HuH-28','HuCCT-1','KKU-100','YSCCC','KKU-055','TFK-1','SSP-25','KKK-D068')

dat.expr$CCA.subgroup <- ifelse(dat.expr$Cell %in% S1,'Subgroup1','Subgroup2')

# subset only genes identified as biomarkers for CCA subtypes
genes.leadingedge <- BB %>% subset(Drug == 'Subtype_1.2')
genes.biomarker <- Biomarkers.test.list %>% subset(Drug == 'Subtype_1.2' & Varimp.Score > 80)

# list of biomarkers
my.genes <- intersect(genes.biomarker$Gene, genes.leadingedge$Gene)

dat.expr2 <- dat.expr %>% subset(Gene %in% unique(my.genes))

# reshape data for pheatmap
datsub.expr.mat <- reshape2::dcast(dat.expr2, Gene ~ Cell, value.var = 'expr') %>% 
  tibble::column_to_rownames(., var = 'Gene')
```

Generate frequency tables of genes in hallmarks
```{r}
my.leadgenes <- genes.leadingedge %>% dplyr::select('Gene','Hallmark') %>% subset(Gene %in% my.genes)
# create table for freq of genes in different hallmarks
dat.freq <- reshape2::dcast(my.leadgenes, Gene ~ Hallmark) %>% tibble::column_to_rownames(., var = 'Gene')

# dat.freq <- ifelse(is.na(dat.freq), 'N','Y') %>% as.data.frame()

for (n in 1:nrow(dat.freq)) {
  dat.freq[n,][which(dat.freq[n,] != 0)] <- 'Y'
  dat.freq[n,][which(dat.freq[n,] == 0)] <- 'N'
  }

# subset only significantly enriched gene sets
sig.hallmark <- df.input %>% subset(Cell == 'Subtype_1.2' & p.adjust < 0.05)
sig.hallmark <- sig.hallmark$ID

sel.dat.freq <- dat.freq %>% as.data.frame() %>% select_(.dots = c(sig.hallmark))

###Prepare color annotation
# color scale with 'white' as 0
my.colorgradient <- rev(c('#67001f','#b2182b','#d6604d','#f4a582','#ffffff','#92c5de','#4393c3','#2166ac','#053061'))

### Column Annotation
# Drug annotation
ls.newnames <- colnames(datsub.expr.mat)
annotation <- data.frame(ID = ls.newnames)

annotation$Subgroup <- ifelse(annotation$ID %in% S1,'CCA.S1','CCA.S2') %>% as.factor()

annotation <- tibble::column_to_rownames(annotation, var = 'ID')

### Row Annotation
#### Tally hallmarks and ranking
ranking <- lapply(1:ncol(dat.freq), function(n) {
  c <- sum(grepl('Y',dat.freq[,n])) 
  h <- colnames(dat.freq)[n]
  d <- data.frame('Hallmark.ID' = h,
                  'Cumsum' = c)
  return(d)
})
# build a dataframe
ranklist <- do.call(rbind,ranking)
# sort Hallmark.ID by Cumsum
ranklist <- ranklist %>% arrange(desc(Cumsum))
# select only significant gene sets
ranklist <- ranklist %>% subset(Hallmark.ID %in% sig.hallmark)

# rearrange pathway columns by ranklist
annotation.row <- dat.freq[,c(ranklist$Hallmark.ID)] %>% as.data.frame()

colnames(annotation.row) <- gsub('HALLMARK_','',colnames(annotation.row))

# spacify colors
varnames <- gsub('HALLMARK_','',sig.hallmark)
mylist<- setNames(as.list(seq(1,length(varnames))),varnames)

for (i in 1:length(mylist)) {
  mylist[[i]] <- c('N' = 'white','Y' = 'black')
}

mylist$Subgroup <- c('CCA.S1' = 'cyan','CCA.S2' = 'orange')

# anno_colors <- list(Subgroup = c('Migrastatic' = 'blue',
#                                  'Cytotoxic' = 'red'),
#                     TNFA_SIGNALING_VIA_NFKB = c('N' = 'white', 'Y' = 'black'),
#                     EPITHELIAL_MESENCHYMAL_TRANSITION = c('N' = 'white', 'Y' = 'black'),
#                     INFLAMMATORY_RESPONSE = c('N' = 'white', 'Y' = 'black'),
#                     P53_PATHWAY = c('N' = 'white', 'Y' = 'black'),
#                     INTERFERON_GAMMA_RESPONSE = c('N' = 'white', 'Y' = 'black'),
#                     MYOGENESIS = c('N' = 'white', 'Y' = 'black'),
#                     APOPTOSIS = c('N' = 'white', 'Y' = 'black'),
#                     COMPLEMENT = c('N' = 'white', 'Y' = 'black'),
#                     KRAS_SIGNALING_UP = c('N' = 'white', 'Y' = 'black'),
#                     XENOBIOTIC_METABOLISM = c('N' = 'white', 'Y' = 'black'),
#                     ESTROGEN_RESPONSE_EARLY = c('N' = 'white', 'Y' = 'black'),
#                     ESTROGEN_RESPONSE_LATE = c('N' = 'white', 'Y' = 'black')
# )

# extrafont::loadfonts(device = 'win')
# windowsFont('Arial')
###

```


```{r}
pt <- datsub.expr.mat
pt.sd <- lapply(1:nrow(pt), function(n) {
  
  sd(pt[n,])
  
})

pt.sd <- do.call(rbind,pt.sd)

id.0sd.row <- which(pt.sd == 0) 

pt.sds <- lapply(1:ncol(pt), function(n) {
  
  sd(pt[,n])
  
})

pt.sds <- do.call(rbind,pt.sds)

id.0sd.col <- which(pt.sds == 0)

if (is_empty(id.0sd.row)) {
  ppt <- pt
} else {
  ppt <- pt[-c(id.0sd.row),]
}

if (is_empty(id.0sd.col)) {
  ppt2 <- ppt
} else {
  ppt2 <- ppt[,-c(id.0sd.col)]
}
require(corrplot)
require(dendsort)
hc <- hclust(dist(ppt2), method = 'average')
callback <- function(hc, ppt2){dendsort(hc,isReverse = TRUE)}


## For pheatmap_1.0.8 and later:
draw_colnames_90 <- function (coln, gaps, ...) {
    coord = pheatmap:::find_coordinates(length(coln), gaps)
    x = coord$coord - 0.5 * coord$size
    res = textGrob(coln, x = x, y = unit(1, "npc") - unit(3,"bigpts"), vjust = 0.5, hjust = 0, rot = 270, gp = gpar(...))
    return(res)}

## 'Overwrite' default draw_colnames with your own version 
assignInNamespace(x="draw_colnames", value="draw_colnames_90",
ns=asNamespace("pheatmap"))

## Try it out
pheatmap(d)


require(pheatmap)
fig4B <- pheatmap::pheatmap(ppt2, 
                  scale = 'row',       # 'row' 'column' 'none'
                  cluster_rows = TRUE,                    # boolean for cluster by rows
                  cluster_cols = TRUE,                    # boolean for cluster by columns
                  clustering_method = 'average',          # method for clustering
                  clustering_distance_rows = 'euclidean', # use euclidean distance
                  clustering_distance_cols = 'euclidean', # use euclidean distance
                  # clustering_distance_rows = dist((1-cor(t(ppt2), method = 'spearman'))),
                  # clustering_distance_cols = dist((1-cor(ppt2, method = 'spearman'))),
                  # clustering_callback = callback,
                  cellwidth = 10,                         # set cell width
                  cellheight = 2,                       # set cell height
                  treeheight_row = 0,
                  treeheight_col = 20,
                  # #cutree_rows = 4,                        # set break gap between rows
                  # #cutree_cols = 5,                        # set break gap between columns
                  show_colnames = TRUE,                   # boolean for showing column names
                  show_rownames = FALSE,                   # boolean for showing row names
                  annotation_colors = mylist,
                  annotation_col = annotation,
                  annotation_row = annotation.row,
                  # fontsize = 10,
                  color = my.colorgradient
)

```

Save figure 4B
```{r}
# setwd('C:/Users/patipark/Dropbox/CCA project/CCA RNA-seq/NewGeneSet_Module_InitialAnalysis/Figure_4B_files')
# extrafont::loadfonts(device = 'pdf')
# windowsFonts(Arial = windowsFont('Arial'))
# library(grid)
# library(gridExtra)
# ggsave(grid.draw(fig4B), 
#        file = 'ClusterpHeatmap_CCA_SigHallmarks_07042019.png', 
#        h = 20, 
#        w = 20, 
#        units = 'cm', 
#        # device = cairo_pdf, # to save in PDF
#        type = 'cairo',
#        dpi = 600)
# dev.off()
```

